addons/base_vat/models/res_partner.py

```diff
@@ -99,161 +92,141 @@ class ResPartner(models.Model):
-            return False
-        check_func_name = 'check_vat_' + country_code
-        check_func = getattr(self, check_func_name, None) or getattr(stdnum.util.get_cc_module(country_code, 'vat'), 'is_valid', None)
-        if not check_func:
-            # No VAT validation available, default to check that the country code exists
-            country_code = _eu_country_vat_inverse.get(country_code, country_code)
-            return bool(self.env['res.country'].search([('code', '=ilike', country_code)]))
-        return check_func(vat_number)
-
-    @api.depends('vat', 'country_id')
-    def _compute_vies_vat_to_check(self):
-        """ Retrieve the VAT number, if one such exists, to be used when checking against the VIES system """
-        eu_country_codes = self.env.ref('base.europe').country_ids.mapped('code')
-        for partner in self:
-            # Skip checks when only one character is used. Some users like to put '/' or other as VAT to differentiate between
-            # a partner for which they haven't yet input VAT, and one not subject to VAT
-            if not partner.vat or len(partner.vat) == 1:
-                partner.vies_vat_to_check = ''
-                continue
-            country_code, number = partner._split_vat(partner.vat)
-            if not country_code.isalpha() and partner.country_id:
-                country_code = partner.country_id.code
-                number = partner.vat
-            partner.vies_vat_to_check = (
-                country_code.upper() in eu_country_codes or
-                country_code.lower() in _region_specific_vat_codes
-            ) and self._fix_vat_number(country_code + number, partner.country_id.id) or ''
+    def _run_vat_checks(self, country, vat, partner_name='', validation='error'):
+        """
+         Returns vat, country.
+          validation: if vat is not valid, then it raises when validation is error and returns '' when validation is setnull
+          partner_name: is for providing a good error message
+        """
+        if not country:
+            return vat, False
+        if not vat or len(vat) == 1:
+            return vat, False
+        vat_prefix, vat_number = self._split_vat(vat)
+
+        if vat_prefix == 'EU' and country not in self.env.ref('base.europe').country_ids:
+            # Foreign companies that trade with non-enterprises in the EU
+            # may have a VATIN starting with "EU" instead of a country code.
+            return vat, False
+
+        prefixed_country = ''
+        if country in self._get_eu_prefixed_countries() and vat_prefix:
+            country_code = EU_EXTRA_VAT_CODES_INV.get(vat_prefix, vat_prefix)
+            if country_code in self._get_eu_prefixed_countries().mapped('code'):
+                vat = vat_number
+                prefixed_country = vat_prefix
+                split_vat = True
+                while split_vat: # For the case where the user would accidently repeat the prefix e.g. BEBEBE0477472701
```

Comment:
I'm not gonna block for that but I'm not sure about this. The user can also do `EBEBE` and then it won't work. Kind of weird that we try to "fix" that case but not others cases.

(We can discuss this after the freeze if needed.)

Reply:
EB would not be a European country in the first place, so it would (should) fail.